### ISSUE TITLE
wrong behavior when setting item on a PyProxyJsonAdaptor

### ISSUE THREAD

**Issue by CNSeniorious000** [pyodide/pyodide#4824]
## 🐛 Bug

`PyProxyJsonAdaptorDictHandlers`'s `set` method force cast the key to `Number`

<details><summary>Screenshot</summary>
<p>
<img src="https://github.com/pyodide/pyodide/assets/74518716/890d4f12-9f22-4c44-b522-ca97ddc86101">
</p>
</details> 

### To Reproduce

```js
> d = pyodide.runPython("dict()").asJsJson()
Proxy(g) {Symbol(pyproxy.attrs): {…}}
> d.set("a", 123)
undefined
> d["b"] = 234
234
> d.toJs()
Proxy(Map) {'a' => 123, NaN => 234}
```

### Expected behavior

```js
> d.toJs()
Proxy(Map) {'a' => 123, 'b' => 234}
```

### Environment

- Pyodide Version `0.26.0`
- Browser version `Edge 125.0.2535.67`
- Any other relevant information: nope

### Additional context

```ts
  set(jsobj: PyProxy, jskey: any, jsval: any): boolean {
    if (typeof jskey === "string") {
      try {
        PySetItemMethods.prototype.set.call(jsobj, Number(jskey), jsval);
```

I think here is why this happens: [`src/core/pyproxy.ts#L2374-L2376`](https://github.com/pyodide/pyodide/blob/main/src/core/pyproxy.ts#L2374-L2376)

**Comment by CNSeniorious000** [comment]
Additionally, the regex to detect number-like string is incorrect:

`""` matches `/^[0-9]*$/` so it will be treat as `Number("")` which is 0.

I think `/^[0-9]+$/` would fix this maybe.

**Comment by CNSeniorious000** [comment]
Here is the reproduction:

```js
> o = pyodide.runPython("{}").asJsJson()
Proxy(g) {Symbol(pyproxy.attrs): {…}}
> a[""] = 1
1
> a[0]
1
> a.toJs()
Proxy(Map) {0 => 1}
```

Same problem exists on arrays:

```js
> a = pyodide.runPython("[123]")
Proxy(g) {Symbol(pyproxy.attrs): {…}}
> a[""] = 234
234
> a.toJs()
[234]
```

Both problem solved in #4825, I think it is well-prepared for review now :) can somebody take a look?

---

Note that the readthedocs's `console.html` always use the latest dev branch's `pyodide.asm.js`, so do we have a "branch preview" url or something? I now can only preview my change by manually downloading the build artifacts 😂
